Allow command-line arguments for "ddev config", fixes #476

rfay opened the pull request:
## The Problem/Issue/Bug:

OP #476: We need to handle config without user prompting, so we'll add new flags to `ddev config`.

## How this PR Solves The Problem:

* Add new flags for configuration --sitename --docroot --pantheon-environment
* Rename Pantheon's "Environment" to "EnvironmentName", since there was also an "environment" and it was impossible to figure out which was which.
* Rename Config() to PromptForConfig() to make the code more readable.
* Refactor Config() so it doesn't return an error when there's no existing config.yaml - instead it only returns an error when the existing has problems.


## Manual Testing Instructions:

Start with no site .ddev file.
* Test existing config paths (`ddev config`, `ddev config pantheon`
* `ddev config --sitename=junker99`, `ddev config --sitename=junker99 --docroot=docroot`
* `ddev config pantheon --pantheon-environment=dev --sitename=rfay-drupal8` and then try a `ddev pull`


## Automated Testing Overview:

I didn't add any new automated testing.

## Related Issue Link(s):

OP #476

## Release/Deployment notes:
<!-- Does this affect anything else, or are there ramifications for other code? Does anything have to be done on deployment? -->

rfay commented:
This showstoppper issue is ready for review. Note that a couple of refactors were done because they were there and I had to find my way around them.  They're bulleted in the OP and are mostly separate commits.

rickmanelius commented:
Just trying to figure out why some GitHub issues provide a more obvious link to the PR. Test: https://github.com/drud/ddev/issues/476

rickmanelius commented:
Review in flight...

## Reference

* Issue https://github.com/drud/ddev/issues/476
* PR https://github.com/drud/ddev/pull/495

## Preparation

* `git checkout rfay/20171002_ddev_config_flags`
* `make clean && make darwin`
* `ddev version` outputs `v0.9.2-6-ge59635d2`, which matches here https://github.com/drud/ddev/pull/495/commits/e59635d26417ae51eb0cab761f3d7e0022f25065.
* Clean out site `ddev rm --remove-data` and `rm -rf .ddev`

## Testing
* `ddev config --help` displays:

```
--docroot string                Provide the relative docroot of the site, like 'docroot' or 'htdocs' or 'web', defaults to empty, the current directory
-h, --help                          help for config
--pantheon-environment string   Provide the environment for a Pantheon site (Pantheon-only) (default "dev")
--sitename string               Provide the sitename of site to configure (normally the same as the directory name)
```

* `ddev config --sitename=junker99`
  * This fails because a docroot isn't set and the app type isn't detected. See https://github.com/drud/wordpress.

rickmanelius commented:
After a chat w/@rfay, updated summary:

* The type flag added.
* Tests for flag names will be on ddev UI.
* Docs will be re-evaluated in the larger meta-issue here https://github.com/drud/ddev/issues/242

beeradb commented on cmd/ddev/cmd/config.go:
```diff
@@ -33,22 +39,43 @@ var ConfigCommand = &cobra.Command{
-		if err != nil {
-			util.Failed("There was a problem configuring your application: %v\n", err)
-		}
+		if siteName == "" && docrootRelPath == "" && pantheonEnvironment == "" {
```
I have mixed feelings about this approach. I'd really like to see a way for providers to each have their own top-level command and register the flags they need. It doesn't feel like this method scales at all beyond our current single provider. it also consolidates a lot of untested logic into our cobra commands, which is a method we've mostly been moving away from. 

On the other hand, I realize this is a blocking issue that we're trying to move on relatively quickly. I also realize we only have a single provider right now, so over-engineering for the future has its own risk. Especially when we have no immediate plans to work on additional providers.

rfay replied:
I think you've nailed it. You're definitely right that this doesn't scale very well, but that there's no reason to deal with the architecture until the next provider.

tannerjfco commented on cmd/ddev/cmd/config.go:
```diff
@@ -33,22 +39,43 @@ var ConfigCommand = &cobra.Command{
+			appType, err := ddevapp.DetermineAppType(c.Docroot)
+			if err != nil {
+				fullPath, _ := filepath.Abs(c.Docroot)
+				util.Failed("Failed to determine app Type (drupal7/drupal8/wordpress), your docroot may be incorrect - looking in directory %v (full path: %v), err: %v", c.Docroot, fullPath, err)
```
- this could end up being a long message, we might want to consider breaking it to another line.
- i feel like we should avoid displaying "err" to end users. this is already an error message, it seems weird to call out an error inside it. "err" has meaning to go devs, and the implication of "error" is probably enough for most others, but I could see it coming off as a weird message.
- suggestion: "your docroot %v[docroot] may be incorrect - looking in directory %v[abspath]: %v[err]"

rfay replied:
On err: Probably if we're going to get a style guide together for error messages we should get one together, and I fully support that.  
I'm always on the debugger's side of this one; in this particular case the user has used a flag that we don't document or suggest. Implementing your suggestion though, which displays the err to end-users anyway :) Since mostly we're the ones that are going to hear about it we need to get this info (especially the err).

tannerjfco commented on cmd/ddev/cmd/config.go:
```diff
@@ -33,22 +39,43 @@ var ConfigCommand = &cobra.Command{
+			// But pantheon *does* validate "Name"
+			err = prov.ValidateField("Name", c.Name)
+			if err != nil {
+				util.Failed("Failed to validate sitename %v with ddev prov %v err: %v", c.Name, provider, err)
```
Suggestion: "Failed to validate sitename %v[sitename] with %v[provider] provider: %v[err]"

tannerjfco commented on pkg/ddevapp/config.go:
```diff
@@ -102,12 +102,14 @@ func NewConfig(AppRoot string, provider string) (*Config, error) {
+	if _, err := os.Stat(c.ConfigPath); !os.IsNotExist(err) {
+		err = c.Read()
+		if err != nil {
+			return c, fmt.Errorf("config.yaml exists but cannot be successfully read, %v, err=%v", c.ConfigPath, err)
```
Suggestion: "%v[configPath] exists but cannot be read: %v[err]"

rfay commented:
* Added the --apptype flag
* Thanks for the suggested improved error messages @tannerjfco - did those.

tannerjfco commented on pkg/ddevapp/config.go:
```diff
@@ -102,12 +102,14 @@ func NewConfig(AppRoot string, provider string) (*Config, error) {
+	if _, err := os.Stat(c.ConfigPath); !os.IsNotExist(err) {
+		err = c.Read()
+		if err != nil {
+			return c, fmt.Errorf("%v[configPath] exists but cannot be read: %v[err]", c.ConfigPath, err)
```
one of my poorly-formatted suggestions missed the roll-back. sorry for the confusion here.

rfay replied:
Thanks for the sharp eye, fixed.

beeradb commented:
This could probably use some additional validation. Passing in a partial arguments list works as intended but results in some odd messaging. Example:

```
testing/drupal-kickstart ‹minikube› git:(master) » ddev config pantheon --pantheon-environment=dev --docroot=docroot
Found a drupal7 codebase at /Users/beeradb/Projects/testing/drupal-kickstart/docroot
Failed to validate sitename  with provider pantheon: could not find a pantheon site named
```

beeradb commented on cmd/ddev/cmd/config.go:
```diff
@@ -33,22 +47,46 @@ var ConfigCommand = &cobra.Command{
+					fullPath, _ := filepath.Abs(c.Docroot)
+					util.Failed("Failed to determine app type (drupal7/drupal8/wordpress).\nYour docroot %v may be incorrect - looking in directory %v, error=%v", c.Docroot, fullPath, err)
+				}
+			}
```
We probably still want to validate appType against the docroot provided and give a warning if there is a mismatch.

Here's an example of what happened when I used "wordpress" as the apptype for a d7 site:
```
testing/drupal-kickstart ‹minikube› git:(master) » ddev config pantheon --pantheon-environment=dev --docroot=docroot --apptype=wordpress --sitename=d7kickstart
Found a drupal7 codebase at /Users/beeradb/Projects/testing/drupal-kickstart/docroot
Configuration complete. You may now run `ddev start` or `ddev pull`
Initial configuration file written successfully. See the configuration file for additional configuration options.
```

rfay replied:
This looks correct to me. If you specify an appType, it's explicitly overriding the auto-detect. What should it do instead of that?  

What kind of generic validation of the docroot would you recommend? 

Essentially the command-line is overrides "I know what I'm doing, shut up now".  If that's not what we want, something else can happen, but it looks to me like that's what you have here.

beeradb replied:
An ideal solution would probably be to accept their value but warn them about a mismatch. i.e. some yellow warning text (or similar) that says something like:

"You specified WordPress as the site type, but this appears to be a Drupal site".

Even people that know what they are doing make mistakes! :)

beeradb replied:
> What kind of generic validation of the docroot would you recommend?

Again, probably fine to accept the value, but a warning if the directory doesn't exist might be nice. Just in case someone makes a mistake.

beeradb commented:
This seems to break regular ddev config behavior. Example:

```
testing/drupal-kickstart ‹minikube› git:(master) ✗ » ddev config
Failed to determine app type (drupal7/drupal8/wordpress).
Your docroot  may be incorrect - looking in directory /Users/beeradb/Projects/testing/drupal-kickstart, error=DetermineAppType() couldn't determine app's type
```

beeradb commented on cmd/ddev/cmd/config.go:
```diff
@@ -56,12 +58,33 @@ var ConfigCommand = &cobra.Command{
+			if provider != "pantheon" && pantheonEnvironment != "" {
+				util.Failed("--pantheon-environment can only be used with pantheon provider, for example ddev config pantheon --pantheon-environment=dev --docroot=docroot")
+			}
+			if pantheonEnvironment != "" && pantheonEnvironment != "dev" && pantheonEnvironment != "test" && pantheonEnvironment != "prod" {
```
Small nit: in multidev environments a different value is completely valid. The pantheon plugin will validate this and the sitename if you use `provider.Validate()` I believe.

rfay commented:
Thanks for the great bug catch and other coaching, @beeradb - The bug was quite simple, but the fixes to validation are adding more spaghetti than I'd like. But it's real, and it's much more validated. Docroot validation is "does it exist". Sitename and environment are actually validated via pantheon. apptype is forced to d7/d8/wordpress.

rfay commented:
Rebased to current upstream/master

rickmanelius commented:
What's weird is that I ran across the similar issue with core `ddev config` behaving badly, but I glossed over it given the tests passing. Given that @beeradb hit it, we probably need to manually step through that a few times to make sure we didn't miss a use case. Sorry about the false positive, @rfay. I'm happy to address the manual review again once the items @beeradb surfaced have been implemented. Thanks!

rickmanelius commented:
Reset the reviewers.

rickmanelius commented:
Good catches, @beeradb. Summarizing:
* Optimization: We may need a test to catch the regression of `sitename` resetting.
* Optimization: Provide warning for app type mismatch (only occurs with flags)
* Consistency: Ensure `config pantheon` reports similar messaging when updating config.

All 3 are valid/valuable. I believe the first two could prevent user errors whereas the 3rd would be a nice to have.

rfay commented:
You're a rockstar catching those issues @beeradb . I've addressed them each in separate commits. 

Essentially this add-on code for config stinks pretty badly, should have been developed in tandem with the related PromptForConfig() logic instead of being tacked in later. But that's where we're at.  It's hard to justify a significant rework for obscure stuff like this, also hard to justify so much spaghetti code for a simple feature. Hopefully bugs that remain will affect only users of the flags, who hopefully will be only ddev-ui, which will use them predictably.

There is a test issue that can be addressed. In 94fe802 I removed 2 (identical) lines of test code that looked for explicit "Creating a new ddev project", which is currently output by util.Success() and via the color output gets out somehow to a different stdout instance. I can either 

1. Switch back to fmt.Printf() to output that
2. Remove the test for that exact phrase (as I did in 94fe802) or 
3. Figure out how to capture the stdout that color package is using (which seems quite extravagant)

rickmanelius commented:
@beeradb I think my review of a +1 holds true as long as this addresses the items summarized here https://github.com/drud/ddev/pull/495#issuecomment-335557913. Any chance you'd be able to test/confirm today?

rfay commented:
IMO this can go in *after* we cut a v0.9.3, and that will relieve any worries about unanticipated behavior, and it can also provide @andrew-c-tran with what he needs.

beeradb commented on cmd/ddev/cmd/config.go:
```diff
@@ -33,22 +52,81 @@ var ConfigCommand = &cobra.Command{
+
+			// c.Name gets set to basename if not provided, or set to sitneName if provided
+			if c.Name != "" && siteName == "" { // If we already have a c.Name and no siteName, leave c.Name alone
+				// Sorry this is empty but it make the logic clearer.
```
nit: "but it make*s* the logic clearer"

rfay replied:
I can handle nits like that!
